R/vcov3_fixest.R

```diff
@@ -244,3 +263,113 @@ vcov_CR3J.fixest <- function(
   res
 }
 
+calculate_beta_jack_sparse = function(obj, cluster, type, absorb_cluster_fixef, return_all) {
+
+  X <- fixest::sparse_model_matrix(obj, type = c("rhs", "fixef"), collin.rm = TRUE)
+  y <- model.matrix(obj, type = "lhs")
+
+  N <- nrow(X)
+  # k: see below
+  w <- weights(obj)
+
+  if(!is.null(w)){
+    X <- sqrt(w) * X
+    y <- sqrt(w) * y
+  }
+
+  # get the clustering variable
+  if(!inherits(cluster, "formula")){
+    cluster <- reformulate(cluster)
+  }
+  cluster_vars = attr(terms(cluster), "term.labels")
+  
+  # Grabs data from the fixest object
+  data = fetch_data(obj, "To apply 'sparse_model_matrix', ")
+
+  # Check that cluster vars are in the original estimation dataset
+  cluster_vars_in_data = cluster_vars %in% colnames(data)
+  if (any(!(cluster_vars_in_data))) {
+    stop(paste0(
+      "The following variables are not found in the dataset used in your `feols` call: ", 
+      paste(cluster_vars[!cluster_vars_in_data], collapse = ", ")
+    ))
+  }
+
+  # Assumes that length(cluster_vars) == 1 (for now; can modify later for multi-way clustering) 
+  if (length(cluster_vars) > 1) stop("Only 1 cluster variable supported right now")
+
+  cluster_vec = data[[cluster_vars]]
```

Comment:
This causes some issues if there are NA variables in the cluster vector. 

E.g. if the first element of `cluster_vars` is not dropped, `unique_clusters` below will have a cluster group "NA". 

In my understanding, fixest handles this by already dropping columns from y and X if there is a NA value in the clustering variable. If the clustering variable is changed on the fly, I think that fixest throws an error. This is certainly what sandwich does. One solution here would be to enforce that 

```r
length(cluster_vec) == nrow(X)
```
and require the user to clean the clustering variable prior to using `vcov_CR3J`.

Reply:
That's actually something I really need to fix for the current version of `summclust`, I will hopefully release a new version today.

Reply:
[This ](https://github.com/s3alfisc/summclust/blob/0f99aaf2846f94ea67b05440528fa6ceb4d57e9a/R/get_cluster.R#L81)is how you can drop missings dropped by `fixest`. Then just throw an error if the cluster vector is still not of the same dimension as the model matrix / if it still includes errors.

Reply:
Yeah, I probably should have been more careful about that. Keeping only the rows used in the original estimation, `cluster_var[object$obs_selection]`, makes sense to me. 

I think that if there are any NAs in `cluster_var[object$obs_selection]`, then an error should be returned. I think that b/c the point estimates will change relative to the original regression and that can be confusing.

Reply:
Yes, I think that's more or less the standard - sandwich, fwildclusterboot and fixest all do this. The released version of summclust will hopefully also do this soon 😀